src/Factory/RepositoryContainer.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Cycle\Factory;
+
+use Cycle\ORM\ORMInterface;
+use Cycle\ORM\RepositoryInterface;
+use Cycle\ORM\SchemaInterface;
+use Psr\Container\ContainerInterface;
+use Yiisoft\Definitions\Exception\NotFoundException;
+use Yiisoft\Definitions\Exception\NotInstantiableClassException;
+
+use function is_string;
+
+final class RepositoryContainer implements ContainerInterface
+{
+    private ContainerInterface $rootContainer;
+    private ORMInterface $orm;
+
+    private bool $rolesBuilt = false;
+    private array $roles = [];
+
+    private array $instances;
+
+    public function __construct(ContainerInterface $rootContainer)
+    {
+        $this->rootContainer = $rootContainer;
+    }
+
+    public function get($id)
+    {
+        if (isset($this->instances[$id])) {
+            return $this->instances[$id];
+        }
+
+        if ($this->has($id)) {
+            return $this->instances[$id] = $this->makeRepository($this->roles[$id]);
+        }
+
+        throw new NotFoundException($id);
+    }
+
+    public function has($id): bool
+    {
+        if (!is_subclass_of($id, RepositoryInterface::class)) {
+            throw new NotInstantiableClassException(
+                $id,
+                sprintf('Can not instantiate "%s" because it is not a subclass of "%s".', $id, RepositoryInterface::class)
+            );
```

Comment:
Are you sure that here you shouldn't return false instead of exception thrown?
Or is it CompositeContainer workflow feature?

Reply:
Already did that.